core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -249,50 +249,70 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
     }
 
     /**
-     * Compiles a script straight-away, or returns the previously compiled and cached script, without checking if it can be executed based on settings.
+     * Compiles a script straight-away, or returns the previously compiled and cached script,
+     * without checking if it can be executed based on settings.
      */
     public CompiledScript compileInternal(Script script) {
         if (script == null) {
             throw new IllegalArgumentException("The parameter script (Script) must not be null.");
         }
 
-        String lang = script.getLang();
+        String lang = script.getLang() == null ? defaultLang : script.getLang();
+        ScriptType type = script.getType();
+        //script.getScript() could return either a name or code for a script,
+        //but we check for a file script name first and an indexed script name second
+        String name = script.getScript();
 
-        if (lang == null) {
-            lang = defaultLang;
-        }
         if (logger.isTraceEnabled()) {
-            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, script.getType(), script.getScript());
+            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, type, name);
         }
 
         ScriptEngineService scriptEngineService = getScriptEngineServiceForLang(lang);
-        String cacheKey = getCacheKey(scriptEngineService, script.getScript());
+        String cacheKey;
+
+        CompiledScript compiledScript;
 
         if (script.getType() == ScriptType.FILE) {
-            CompiledScript compiled = staticCache.get(cacheKey); //On disk scripts will be loaded into the staticCache by the listener
-            if (compiled == null) {
-                throw new IllegalArgumentException("Unable to find on disk script " + script.getScript());
+            cacheKey = getCacheKey(scriptEngineService, name, null);
+            //On disk scripts will be loaded into the staticCache by the listener
+            compiledScript = staticCache.get(cacheKey);
+
+            if (compiledScript == null) {
+                compiledScript = new CompiledScript(type, name, lang, null);
+                throw new IllegalArgumentException("Unable to find on disk " + compiledScript);
             }
-            return compiled;
+
+            return compiledScript;
         }
 
+        //script.getScript() will be code if the script type is inline
         String code = script.getScript();
 
         if (script.getType() == ScriptType.INDEXED) {
-            final IndexedScript indexedScript = new IndexedScript(lang, script.getScript());
+            final IndexedScript indexedScript = new IndexedScript(lang, name);
```

Comment:
maybe add a comment here that we must lookup the indexed script before the cache because the code may have changed?